src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+
+    const Register haystack     = rdi;
+    const Register haystack_len = rsi;
+    const Register needle       = rdx;
+    const Register needle_len   = rcx;
+
+    const Register save_ndl_len = r12;
+
+    const XMMRegister byte_0    = XMM_BYTE_0;
+    const XMMRegister byte_k    = XMM_BYTE_K;
+    const XMMRegister byte_1    = XMM_BYTE_1;
+    const XMMRegister byte_2    = XMM_BYTE_2;
+    const XMMRegister xmm_tmp1  = XMM_TMP1;
+    const XMMRegister xmm_tmp2  = XMM_TMP2;
+    const XMMRegister xmm_tmp3  = XMM_TMP3;
+    const XMMRegister xmm_tmp4  = XMM_TMP4;
+
+    const XMMRegister save_r12  = xmm4;
+    const XMMRegister save_r13  = xmm5;
+    const XMMRegister save_r14  = xmm6;
+    const XMMRegister save_r15  = xmm7;
+    const XMMRegister save_rbx  = xmm8;
+    // xmm registers more valuable in inner loops...
+    // const XMMRegister save_rsi  = xmm9;
+    // const XMMRegister save_rdi  = xmm10;
+    // const XMMRegister save_rcx  = xmm11;
+    // const XMMRegister save_r8   = xmm12;
+
+    // Addresses of the two jump tables used for small needle processing
+    address jump_table;
+    address jump_table_1;
+
+    Label L_begin;
+
+    Label L_returnRBP, L_checkRangeAndReturn, L_returnError;
+    Label L_bigCaseFixupAndReturn, L_checkRangeAndReturnRCX;
+    Label L_returnZero, L_copyHaystackToStackDone, L_bigSwitchTop;
+    Label L_bigCaseDefault, L_smallCaseDefault, L_copyHaystackToStack, L_smallSwitchTop;
+    Label L_nextCheck, L_checksPassed, L_zeroCheckFailed;
+
+    Label L_wcharBegin, L_continue, L_wideNoExpand, L_copyDone, L_copyHigh;
+    Label L_wideMidLoop, L_wideTopLoop, L_wideInnerLoop, L_wideFound;
+
+    __ align(CodeEntryAlignment);
+    fnptrs[isLL   ? StrIntrinsicNode::LL
+           : isUU ? StrIntrinsicNode::UU
+                  : StrIntrinsicNode::UL] = __ pc();
+    __ enter();  // required for proper stackwalking of RuntimeStub frame
+
+    // Check for trivial cases
+    // needle length == 0?
+    __ cmpq(needle_len, 0);
+    __ jg_b(L_nextCheck);
+    __ xorq(rax, rax);
+    __ leave();
+    __ ret(0);
+
+    // haystack length == 0?
+    __ bind(L_nextCheck);
+    __ testq(haystack_len, haystack_len);
+    __ je(L_zeroCheckFailed);
+
+    // haystack length >= needle length?
+    __ movq(rax, haystack_len);
+    __ subq(rax, needle_len);
+    __ jge_b(L_checksPassed);
+
+    __ bind(L_zeroCheckFailed);
+    __ movq(rax, -1);
+    __ leave();
+    __ ret(0);
+
+    __ bind(L_checksPassed);
+
+    // Check for highly-optimized ability - haystack <= 32 bytes and needle <= OPT_NEEDLE_SIZE_MAX
+    // haystack_len is in elements, not bytes, for UTF-16
+    __ cmpq(haystack_len, isU ? COPIED_HAYSTACK_STACK_SIZE / 4
+                              : COPIED_HAYSTACK_STACK_SIZE / 2);
+    __ ja(L_begin);
+
+    // needle_len is in elements, not bytes, for UTF-16
+    __ cmpq(needle_len, isUU ? OPT_NEEDLE_SIZE_MAX / 2 : OPT_NEEDLE_SIZE_MAX);
```

Comment:
OPT_NEEDLE_SIZE_MAX is an odd number (set to 5), should that have been an even number?